File: side-projects/peri-sync/server/src/limits.ts
/// <reference types="@cloudflare/workers-types" />
// 限额与 TTL 常量（r2-encrypted-transfer contract）。
//
// 数值镜像 Rust `peri-tui/src/sync/limits.rs`（Slice 0 冻结，此后仅经显式
// plan 修订）。部署默认值即下方冻结值；env 覆盖仅用于本地测试推进 alarm
// （03-plan.md §1.6 "env 配置门禁"），生产部署不得覆盖。
//
// 未冻结项（03-plan 未给数值）：非 lookup/code 的签名端点统一采用宽松
// 60/min/device 防滥用，部署门禁可收紧。

export interface RateLimits {
	/** code 注册限流（次/分钟/device）。冻结 2。 */
	codeRegisterMaxPerMin: number;
	/** code lookup 限流（次/分钟/IP）。冻结 15。 */
	codeLookupRateLimitPerMin: number;
	/** 其他签名端点宽松限流（次/分钟/device）。未冻结，部署门禁可调。 */
	signedEndpointRateLimitPerMin: number;
	/** create 端点限流（次/分钟/device）。未冻结，部署门禁可调。 */
	createRateLimitPerMin: number;
}

export interface TtlLimits {
	/** created 状态 TTL（秒）。冻结 600。 */
	createdSecs: number;
	/** paired（join 成功/握手进行中）TTL（秒）。冻结 300。 */
	pairedSecs: number;
	/** ready/transferring TTL（秒）。冻结 3600。 */
	readySecs: number;
	/** 终态 tombstone（秒）。冻结 3600。 */
	tombstoneSecs: number;
}

export interface Limits extends RateLimits, TtlLimits {
	/** 签名时间戳可接受偏差（±300 秒，服务端侧）。 */
	signatureSkewSecs: number;
	/** 单个 payload part（AEAD envelope）上限（64 KiB）。 */
	maxPartBytes: number;
	/** 每 channel 最大 part 数（512）。 */
	maxPartsPerChannel: number;
	/**
	 * 每 channel payload 总预算：64 KiB × 512 = 32 MiB（实测对齐，
	 * L2 修订）；Slice 3 Rust 客户端按 32 MiB 分片，两侧一致。
	 */
	maxPayloadBytes: number;
	/** 单条 handshake opaque 消息上限（4 KiB，Noise blob 富余）。 */
	maxMsgBytes: number;
	/** 入口统一 body 大小上限（128 KiB，覆盖 b64 编码后的 part）。 */
	maxBodyBytes: number;
	/** code 自注册时刻起有效（秒）。冻结 60。 */
	codeValidSecs: number;
}

function readInt(env: Record<string, unknown> | undefined, key: string, fallback: number): number {
	if (!env) return fallback;
	const raw = env[key];
	if (typeof raw === "string" && /^\d+$/.test(raw)) return parseInt(raw, 10);
	if (typeof raw === "number" && Number.isFinite(raw) && raw >= 0) return raw;
	return fallback;
}

export function limits(env?: Record<string, unknown>): Limits {
	return {
		createdSecs: readInt(env, "TTL_CREATED_SECS", 600),
		pairedSecs: readInt(env, "TTL_PAIRED_SECS", 300),
		readySecs: readInt(env, "TTL_READY_SECS", 3600),
		tombstoneSecs: readInt(env, "TTL_TOMBSTONE_SECS", 3600),
		codeValidSecs: readInt(env, "CODE_VALID_SECS", 60),
		signatureSkewSecs: readInt(env, "SIGNATURE_SKEW_SECS", 300),
		maxPartBytes: readInt(env, "MAX_PART_BYTES", 64 * 1024),
		maxPartsPerChannel: readInt(env, "MAX_PARTS_PER_CHANNEL", 512),
		maxPayloadBytes: readInt(env, "MAX_PAYLOAD_BYTES", 32 * 1024 * 1024),
		maxMsgBytes: readInt(env, "MAX_MSG_BYTES", 4 * 1024),
		maxBodyBytes: readInt(env, "MAX_BODY_BYTES", 128 * 1024),
		codeRegisterMaxPerMin: readInt(env, "CODE_REGISTER_MAX_PER_MIN", 2),
		codeLookupRateLimitPerMin: readInt(env, "CODE_LOOKUP_RATE_LIMIT_PER_MIN", 15),
		signedEndpointRateLimitPerMin: readInt(env, "SIGNED_ENDPOINT_RATE_PER_MIN", 60),
		createRateLimitPerMin: readInt(env, "CREATE_RATE_PER_MIN", 60),
	};
}

File: peri-tui/src/sync/limits.rs
//! 客户端侧限额与 TTL 常量（r2-encrypted-transfer contract）。
//!
//! 本模块是客户端预算检查的事实源；Slice 2 的 TS Worker 必须镜像同一组数值
//! （channel TTL、限流、签名偏差与 payload 限额均已在 plan 中冻结）。

use anyhow::Result;

/// 单个 payload part 上限（64 KiB，**密文**口径；TS `maxPartBytes` 同值）。
pub const MAX_PART_BYTES: usize = 64 * 1024;

/// 单个明文分片上限：`MAX_PART_BYTES - ENVELOPE_HEADER_LEN - AEAD_TAG_LEN`
/// （= 65507）。seal 后每片密文恰为 `MAX_PART_BYTES`，与 TS channel-do 的
/// `ct.length > maxPartBytes → 413` 逐字节对齐（C1 复审修复）。
pub const MAX_PLAINTEXT_PART_BYTES: usize =
    MAX_PART_BYTES - crate::sync::crypto::ENVELOPE_HEADER_LEN - crate::sync::crypto::AEAD_TAG_LEN;

/// 每 channel 最大 part 数。
pub const MAX_PARTS_PER_CHANNEL: usize = 512;

/// 每 channel payload 总预算（32 MiB = 64 KiB × 512，L2 修订；与 TS
/// `side-projects/peri-sync/server/src/limits.ts` 的 `MAX_PAYLOAD_BYTES` 一致）。
pub const MAX_PAYLOAD_BYTES: usize = 32 * 1024 * 1024;

/// manifest 上限（256 KiB）。
pub const MAX_MANIFEST_BYTES: usize = 256 * 1024;

/// 设备名最大字符数。
pub const MAX_DEVICE_NAME_CHARS: usize = 64;

/// 签名时间戳可接受偏差（±300 秒，服务端侧）。
pub const SIGNATURE_SKEW_SECS: u64 = 300;

/// channel TTL：created 起 +10 分钟。
pub const TTL_CREATED_SECS: u64 = 600;
/// channel TTL：join 成功/握手进行中 +5 分钟。
pub const TTL_JOINED_SECS: u64 = 300;
/// channel TTL：ready 起 +1 小时。
pub const TTL_READY_SECS: u64 = 3600;
/// 终态 tombstone +1 小时。
pub const TTL_TOMBSTONE_SECS: u64 = 3600;

/// 正常 sender 码注册频率上限（次/分钟）。
pub const CODE_REGISTER_MAX_PER_MIN: u32 = 2;
/// 撞码最多一次额外重试。
pub const CODE_COLLISION_MAX_RETRIES: u32 = 1;
/// code lookup 限流（次/分钟）；避免成功路径被错误 429。
pub const CODE_LOOKUP_RATE_LIMIT_PER_MIN: u32 = 15;

/// 客户端指数退避基准（毫秒）。
pub const BASE_BACKOFF_MS: u64 = 500;
/// 客户端指数退避上限（毫秒）。
pub const MAX_BACKOFF_MS: u64 = 30_000;

/// 校验 manifest 预算（part 数与**密文**总字节），在 R2 PUT 前检查。
///
/// `total_bytes` 为各 part envelope 长度之和，与 TS channel-do 的
/// `total_bytes + ct.length > maxPayloadBytes → 413` 累计口径一致（C1）。
pub fn validate_manifest(parts: usize, total_bytes: usize) -> Result<()> {
    if parts > MAX_PARTS_PER_CHANNEL {
        anyhow::bail!("manifest exceeds part limit: {parts} > {MAX_PARTS_PER_CHANNEL}");
    }
    if total_bytes > MAX_PAYLOAD_BYTES {
        anyhow::bail!("manifest exceeds payload budget: {total_bytes} > {MAX_PAYLOAD_BYTES}");
    }
    Ok(())
}

/// 校验单个 part 大小。
pub fn validate_part_size(part_bytes: usize) -> Result<()> {
    if part_bytes > MAX_PART_BYTES {
        anyhow::bail!("part exceeds size limit: {part_bytes} > {MAX_PART_BYTES}");
    }
    Ok(())
}

/// 校验设备名（非空、不超过 [`MAX_DEVICE_NAME_CHARS`] 字符）。
pub fn validate_device_name(name: &str) -> Result<()> {
    if name.is_empty() {
        anyhow::bail!("device name must not be empty");
    }
    let chars = name.chars().count();
    if chars > MAX_DEVICE_NAME_CHARS {
        anyhow::bail!("device name too long: {chars} chars > {MAX_DEVICE_NAME_CHARS}");
    }
    Ok(())
}
